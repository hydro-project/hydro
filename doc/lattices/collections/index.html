<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>