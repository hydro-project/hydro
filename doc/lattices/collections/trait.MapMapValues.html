<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Trait for transforming the values of a map without changing the overall type of the data structure."><title>MapMapValues in lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-e56847b5.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (f6092f224 2025-09-22)" data-channel="nightly" data-search-js="search-e256b49e.js" data-stringdex-js="stringdex-061df703.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../../static.files/main-6dc2a7f3.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">MapMapValues</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">MapMap<wbr>Values</a></h2><h3><a href="#required-associated-types">Required Associated Types</a></h3><ul class="block"><li><a href="#associatedtype.MapValue" title="MapValue">MapValue</a></li></ul><h3><a href="#required-methods">Required Methods</a></h3><ul class="block"><li><a href="#tymethod.map_values" title="map_values">map_values</a></li></ul><h3><a href="#foreign-impls">Implementations on Foreign Types</a></h3><ul class="block"><li><a href="#impl-MapMapValues%3COldVal%3E-for-BTreeMap%3CKey,+OldVal%3E" title="BTreeMap&#60;Key, OldVal&#62;">BTreeMap&#60;Key, OldVal&#62;</a></li><li><a href="#impl-MapMapValues%3COldVal%3E-for-HashMap%3CKey,+OldVal%3E" title="HashMap&#60;Key, OldVal&#62;">HashMap&#60;Key, OldVal&#62;</a></li></ul><h3><a href="#dyn-compatibility">Dyn Compatibility</a></h3><h3><a href="#implementors">Implementors</a></h3></section><div id="rustdoc-modnav"><h2><a href="index.html">In lattices::<wbr>collections</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a>::<wbr><a href="index.html">collections</a></div><h1>Trait <span class="trait">MapMap<wbr>Values</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#16-24">Source</a> </span></div><pre class="rust item-decl"><code>pub trait MapMapValues&lt;OldVal&gt; {
    type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt;;

    // Required method
    fn <a href="#tymethod.map_values" class="fn">map_values</a>&lt;NewVal, MapFn&gt;(self, map_fn: MapFn) -&gt; Self::<a class="associatedtype" href="trait.MapMapValues.html#associatedtype.MapValue" title="type lattices::collections::MapMapValues::MapValue">MapValue</a>&lt;NewVal&gt;
       <span class="where">where MapFn: <a class="trait" href="https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html" title="trait core::ops::function::FnMut">FnMut</a>(OldVal) -&gt; NewVal</span>;
}</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Trait for transforming the values of a map without changing the overall type of the data structure.</p>
</div></details><h2 id="required-associated-types" class="section-header">Required Associated Types<a href="#required-associated-types" class="anchor">§</a></h2><div class="methods"><details class="toggle" open><summary><section id="associatedtype.MapValue" class="method"><a class="src rightside" href="../../src/lattices/collections.rs.html#18">Source</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt;</h4></section></summary><div class="docblock"><p>Output type, should be <code>Self</code> but with <code>OldVal</code> replaced with <code>NewVal</code>.</p>
</div></details></div><h2 id="required-methods" class="section-header">Required Methods<a href="#required-methods" class="anchor">§</a></h2><div class="methods"><details class="toggle method-toggle" open><summary><section id="tymethod.map_values" class="method"><a class="src rightside" href="../../src/lattices/collections.rs.html#21-23">Source</a><h4 class="code-header">fn <a href="#tymethod.map_values" class="fn">map_values</a>&lt;NewVal, MapFn&gt;(self, map_fn: MapFn) -&gt; Self::<a class="associatedtype" href="trait.MapMapValues.html#associatedtype.MapValue" title="type lattices::collections::MapMapValues::MapValue">MapValue</a>&lt;NewVal&gt;<div class="where">where
    MapFn: <a class="trait" href="https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html" title="trait core::ops::function::FnMut">FnMut</a>(OldVal) -&gt; NewVal,</div></h4></section></summary><div class="docblock"><p>Map the values into using the <code>map_fn</code>.</p>
</div></details></div><h2 id="dyn-compatibility" class="section-header">Dyn Compatibility<a href="#dyn-compatibility" class="anchor">§</a></h2><div class="dyn-compatibility-info"><p>This trait is <b>not</b> <a href="https://doc.rust-lang.org/nightly/reference/items/traits.html#dyn-compatibility">dyn compatible</a>.</p><p><i>In older versions of Rust, dyn compatibility was called "object safety", so this trait is not object safe.</i></p></div><h2 id="foreign-impls" class="section-header">Implementations on Foreign Types<a href="#foreign-impls" class="anchor">§</a></h2><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-BTreeMap%3CKey,+OldVal%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#40-54">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-BTreeMap%3CKey,+OldVal%3E" class="anchor">§</a><h3 class="code-header">impl&lt;Key, OldVal&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/collections/btree/map/struct.BTreeMap.html" title="struct alloc::collections::btree::map::BTreeMap">BTreeMap</a>&lt;Key, OldVal&gt;<div class="where">where
    Key: <a class="trait" href="https://doc.rust-lang.org/nightly/core/cmp/trait.Eq.html" title="trait core::cmp::Eq">Eq</a> + <a class="trait" href="https://doc.rust-lang.org/nightly/core/cmp/trait.Ord.html" title="trait core::cmp::Ord">Ord</a>,</div></h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-1" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#44">Source</a><a href="#associatedtype.MapValue-1" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/collections/btree/map/struct.BTreeMap.html" title="struct alloc::collections::btree::map::BTreeMap">BTreeMap</a>&lt;Key, NewVal&gt;</h4></section><section id="method.map_values" class="method trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#46-53">Source</a><a href="#method.map_values" class="anchor">§</a><h4 class="code-header">fn <a href="#tymethod.map_values" class="fn">map_values</a>&lt;NewVal, MapFn&gt;(self, map_fn: MapFn) -&gt; Self::<a class="associatedtype" href="trait.MapMapValues.html#associatedtype.MapValue" title="type lattices::collections::MapMapValues::MapValue">MapValue</a>&lt;NewVal&gt;<div class="where">where
    MapFn: <a class="trait" href="https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html" title="trait core::ops::function::FnMut">FnMut</a>(OldVal) -&gt; NewVal,</div></h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-HashMap%3CKey,+OldVal%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#25-39">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-HashMap%3CKey,+OldVal%3E" class="anchor">§</a><h3 class="code-header">impl&lt;Key, OldVal&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="https://doc.rust-lang.org/nightly/std/collections/hash/map/struct.HashMap.html" title="struct std::collections::hash::map::HashMap">HashMap</a>&lt;Key, OldVal&gt;<div class="where">where
    Key: <a class="trait" href="https://doc.rust-lang.org/nightly/core/cmp/trait.Eq.html" title="trait core::cmp::Eq">Eq</a> + <a class="trait" href="https://doc.rust-lang.org/nightly/core/hash/trait.Hash.html" title="trait core::hash::Hash">Hash</a>,</div></h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-2" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#29">Source</a><a href="#associatedtype.MapValue-2" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="https://doc.rust-lang.org/nightly/std/collections/hash/map/struct.HashMap.html" title="struct std::collections::hash::map::HashMap">HashMap</a>&lt;Key, NewVal&gt;</h4></section><section id="method.map_values-1" class="method trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#31-38">Source</a><a href="#method.map_values-1" class="anchor">§</a><h4 class="code-header">fn <a href="#tymethod.map_values" class="fn">map_values</a>&lt;NewVal, MapFn&gt;(self, map_fn: MapFn) -&gt; Self::<a class="associatedtype" href="trait.MapMapValues.html#associatedtype.MapValue" title="type lattices::collections::MapMapValues::MapValue">MapValue</a>&lt;NewVal&gt;<div class="where">where
    MapFn: <a class="trait" href="https://doc.rust-lang.org/nightly/core/ops/function/trait.FnMut.html" title="trait core::ops::function::FnMut">FnMut</a>(OldVal) -&gt; NewVal,</div></h4></section></div></details><h2 id="implementors" class="section-header">Implementors<a href="#implementors" class="anchor">§</a></h2><div id="implementors-list"><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-OptionMap%3CK,+OldVal%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#951-960">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-OptionMap%3CK,+OldVal%3E" class="anchor">§</a><h3 class="code-header">impl&lt;K, OldVal&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">OptionMap</a>&lt;K, OldVal&gt;</h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-3" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#952">Source</a><a href="#associatedtype.MapValue-3" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">OptionMap</a>&lt;K, NewVal&gt;</h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-SingletonMap%3CK,+OldVal%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#702-713">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-SingletonMap%3CK,+OldVal%3E" class="anchor">§</a><h3 class="code-header">impl&lt;K, OldVal&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">SingletonMap</a>&lt;K, OldVal&gt;</h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-4" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#703">Source</a><a href="#associatedtype.MapValue-4" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">SingletonMap</a>&lt;K, NewVal&gt;</h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-VecMap%3CK,+OldVal%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#283-294">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-VecMap%3CK,+OldVal%3E" class="anchor">§</a><h3 class="code-header">impl&lt;K, OldVal&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a>&lt;K, OldVal&gt;</h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-5" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#284">Source</a><a href="#associatedtype.MapValue-5" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a>&lt;K, NewVal&gt;</h4></section></div></details><details class="toggle implementors-toggle"><summary><section id="impl-MapMapValues%3COldVal%3E-for-ArrayMap%3CK,+OldVal,+N%3E" class="impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#1192-1203">Source</a><a href="#impl-MapMapValues%3COldVal%3E-for-ArrayMap%3CK,+OldVal,+N%3E" class="anchor">§</a><h3 class="code-header">impl&lt;K, OldVal, const N: <a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.usize.html">usize</a>&gt; <a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMapValues</a>&lt;OldVal&gt; for <a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">ArrayMap</a>&lt;K, OldVal, N&gt;</h3></section></summary><div class="impl-items"><section id="associatedtype.MapValue-6" class="associatedtype trait-impl"><a class="src rightside" href="../../src/lattices/collections.rs.html#1193">Source</a><a href="#associatedtype.MapValue-6" class="anchor">§</a><h4 class="code-header">type <a href="#associatedtype.MapValue" class="associatedtype">MapValue</a>&lt;NewVal&gt; = <a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">ArrayMap</a>&lt;K, NewVal, N&gt;</h4></section></div></details></div><script src="../../trait.impl/lattices/collections/trait.MapMapValues.js" data-ignore-extern-crates="std,alloc" async></script></section></div></main></body></html>