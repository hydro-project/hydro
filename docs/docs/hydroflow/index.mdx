import {Redirect} from '@docusaurus/router';

<Redirect to="/docs/dfir" />
