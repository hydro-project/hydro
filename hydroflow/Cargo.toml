[package]
name = "hydroflow"
version = "0.1.0"
edition = "2021"

[features]
default = [ "async", "macros" ]
async = [ "futures" ]
macros = [ "hydroflow_macro", "hydroflow_datalog" ]

[[example]]
name = "echoserver"

[[example]]
name = "echo_serde_json"

[[example]]
name = "chat"

[[example]]
name = "graph_reachability"

[[example]]
name = "graph_unreachability"

[[example]]
name = "three_clique"

[[example]]
name = "two_pc"

[[example]]
name = "deadlock_detector"

[[example]]
name = "kvs"

[dependencies]
bincode = "1.3"
byteorder = "1.4.3"
bytes = "1.1.0"
futures = { version = "0.3", optional = true }
hydroflow_datalog = { optional = true, path = "../hydroflow_datalog" }
hydroflow_lang = { path = "../hydroflow_lang" }
hydroflow_macro = { optional = true, path = "../hydroflow_macro" }
pin-project-lite = "0.2.9"
pusherator = { path = "../pusherator" }
ref-cast = "1.0"
regex = "1"
sealed = "0.4"
serde = { version = "1", features = [ "derive" ] }
serde_json = "1"
slotmap = "1.0"
static_assertions = "1.1.0"
tokio = { version = "1.16", features = [ "full" ] }
tokio-stream = { version = "0.1.10", features = [ "io-util" ] }
tokio-util = { version = "0.7.4", features = [ "net", "codec" ] }
variadics = { path = "../variadics" }

[dev-dependencies]
chrono = { version = "0.4.20", features = [ "serde" ], default-features = false }
clap = { version = "3.0.14", features = [ "derive" ] }
colored = "2.0"
criterion = { version = "0.3", features = [ "async_tokio" ] }
futures = { version = "0.3" }
itertools = "0.10.3"
rand = "0.8.4"
textnonce = "1.0.0"
time = "0.3"
trybuild = "1.0.71"
hdrhistogram = "7"
zipf = "7.0.0"
core_affinity = "0.5.10"
