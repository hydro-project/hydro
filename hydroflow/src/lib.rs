#![feature(never_type)]
#![feature(type_alias_impl_trait)]
#![allow(type_alias_bounds)]
#![allow(clippy::let_and_return)]
#![allow(clippy::iter_with_drain)]
#![allow(clippy::explicit_auto_deref)]

pub mod builder;
pub mod compiled;
pub mod lang;
pub mod props;
pub mod scheduled;

pub use hydroflow_macro::*;

pub use pusherator;
pub use tokio;
pub use tuple_list::tuple_list as tl;
pub use tuple_list::tuple_list_type as tt;

#[cfg(doctest)]
mod booktest {
    macro_rules! booktest {
        ($i:ident) => {
            #[doc = include_str!(concat!("../../book/", stringify!($i), ".md"))]
            mod $i {}
        };
    }
    booktest!(example_1);
    booktest!(example_2);
    booktest!(example_3);
    booktest!(example_4);
    booktest!(example_5);

    booktest!(example_1_surface);
    booktest!(example_2_surface);
    booktest!(example_3_surface);
    booktest!(example_4_surface);
    booktest!(example_5_surface);

    booktest!(surface_syntax);
    booktest!(surface_embedding);
    booktest!(surface_flows);
    booktest!(surface_data);
    booktest!(surface_ops);
}
