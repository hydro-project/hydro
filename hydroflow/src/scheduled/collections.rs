pub struct Single<T>(pub T);

pub struct Iter<I: Iterator>(pub I);
