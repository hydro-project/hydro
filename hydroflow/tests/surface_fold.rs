use std::collections::{HashMap, HashSet};

use hydroflow::{hydroflow_syntax, util::collect_ready};

#[test]
pub fn test_fold_tick() {
    let (items_send, items_recv) = hydroflow::util::unbounded_channel::<Vec<u32>>();
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<Vec<u32>>();

    let mut df = hydroflow::hydroflow_syntax! {
        source_stream(items_recv)
            -> fold::<'tick>(Vec::new(), |mut old: Vec<u32>, mut x: Vec<u32>| { old.append(&mut x); old })
            -> for_each(|v| result_send.send(v).unwrap());
    };

    println!(
        "{}",
        df.serde_graph()
            .expect("No graph found, maybe failed to parse.")
            .to_mermaid()
    );

    items_send.send(vec![1, 2]).unwrap();
    items_send.send(vec![3, 4]).unwrap();
    df.run_available();

    assert_eq!(
        &[vec![1, 2, 3, 4]],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut result_recv)
    );

    items_send.send(vec![5, 6]).unwrap();
    items_send.send(vec![7, 8]).unwrap();
    df.run_available();

    assert_eq!(
        &[vec![5, 6, 7, 8]],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[test]
pub fn test_fold_static() {
    let (items_send, items_recv) = hydroflow::util::unbounded_channel::<Vec<u32>>();
    let (result_send, mut result_recv) = hydroflow::util::unbounded_channel::<Vec<u32>>();

    let mut df = hydroflow::hydroflow_syntax! {
        source_stream(items_recv)
            -> fold::<'static>(Vec::new(), |mut old: Vec<u32>, mut x: Vec<u32>| { old.append(&mut x); old })
            -> for_each(|v| result_send.send(v).unwrap());
    };

    println!(
        "{}",
        df.serde_graph()
            .expect("No graph found, maybe failed to parse.")
            .to_mermaid()
    );

    items_send.send(vec![1, 2]).unwrap();
    items_send.send(vec![3, 4]).unwrap();
    df.run_available();

    assert_eq!(
        &[vec![1, 2, 3, 4]],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut result_recv)
    );

    items_send.send(vec![5, 6]).unwrap();
    items_send.send(vec![7, 8]).unwrap();
    df.run_available();

    assert_eq!(
        &[vec![1, 2, 3, 4, 5, 6, 7, 8]],
        &*hydroflow::util::collect_ready::<Vec<_>, _>(&mut result_recv)
    );
}

#[test]
pub fn test_fold_flatten() {
    // test pull
    let (out_send, mut out_recv) = hydroflow::util::unbounded_channel::<(u8, u8)>();
    let mut df_pull = hydroflow_syntax! {
        source_iter([(1,1), (1,2), (2,3), (2,4)])
            -> fold::<'tick>(HashMap::<u8,u8>::new(), |mut ht, t: (u8,u8)| {
                    let e = ht.entry(t.0).or_insert(0);
                    *e += t.1;
                    ht})
            -> flatten()
            -> for_each(|(k,v)| out_send.send((k,v)).unwrap());
    };
    df_pull.run_available();

    let out: HashSet<_> = collect_ready(&mut out_recv);
    for pair in [(1, 3), (2, 7)] {
        assert!(out.contains(&pair));
    }

    // test push
    let (out_send, mut out_recv) = hydroflow::util::unbounded_channel::<(u8, u8)>();
    let mut df_push = hydroflow_syntax! {
        datagen = source_iter([(1,2), (1,2), (2,4), (2,4)]) -> tee();
        datagen[0] -> fold::<'tick>(HashMap::<u8,u8>::new(), |mut ht, t:(u8,u8)| {
                let e = ht.entry(t.0).or_insert(0);
                *e += t.1;
                ht})
            -> flatten()
            -> for_each(|(k,v)| out_send.send((k,v)).unwrap());
        datagen[1] -> null();
    };

    df_push.run_available();

    let out: HashSet<_> = collect_ready(&mut out_recv);
    for pair in [(1, 4), (2, 8)] {
        assert!(out.contains(&pair));
    }
}

#[test]
pub fn test_fold_sort() {
    let (items_send, items_recv) = hydroflow::util::unbounded_channel::<usize>();

    let mut df = hydroflow_syntax! {
        source_stream(items_recv)
            -> fold::<'tick>(Vec::new(), |mut v, x| {
                v.push(x);
                v
            })
            -> flat_map(|mut vec| { vec.sort(); vec })
            -> for_each(|v| print!("{:?}, ", v));
    };

    println!(
        "{}",
        df.serde_graph()
            .expect("No graph found, maybe failed to parse.")
            .to_mermaid()
    );
    df.run_available();

    print!("\nA: ");

    items_send.send(9).unwrap();
    items_send.send(2).unwrap();
    items_send.send(5).unwrap();
    df.run_available();

    print!("\nB: ");

    items_send.send(9).unwrap();
    items_send.send(5).unwrap();
    items_send.send(2).unwrap();
    items_send.send(0).unwrap();
    items_send.send(3).unwrap();
    df.run_available();

    println!();
}
