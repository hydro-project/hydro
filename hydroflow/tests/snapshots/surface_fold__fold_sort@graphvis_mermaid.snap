---
source: hydroflow/tests/surface_fold.rs
expression: df.meta_graph().unwrap().to_mermaid()
---
%%{init:{'theme':'base','themeVariables':{'clusterBkg':'#ddd','clusterBorder':'#888'}}}%%
flowchart TD
classDef pullClass fill:#8af,stroke:#000,text-align:left,white-space:pre
classDef pushClass fill:#ff8,stroke:#000,text-align:left,white-space:pre
linkStyle default stroke:#aaa,stroke-width:4px,color:red,font-size:1.5em;
subgraph sg_1v1 ["sg_1v1 stratum 0"]
    1v1[\"(1v1) <code>source_stream(items_recv)</code>"/]:::pullClass
end
subgraph sg_2v1 ["sg_2v1 stratum 1"]
    2v1[\"(2v1) <code>fold::&lt;'tick&gt;(Vec::new(), Vec::push)</code>"/]:::pullClass
    3v1[\"<div style=text-align:center>(3v1)</div> <code>flat_map(|mut vec| {<br>    vec.sort();<br>    vec<br>})</code>"/]:::pullClass
    4v1[/"(4v1) <code>for_each(|v| print!(&quot;{:?}, &quot;, v))</code>"\]:::pushClass
    2v1-->3v1
    3v1-->4v1
end
1v1-->5v1
5v1["(5v1) <code>handoff</code>"]:::otherClass
5v1--x2v1

